TRef not accepted in constructor by derive macro

```Rust
#[derive(gdnative::NativeClass)]
#[inherit(gdnative::prelude::Node)]
pub struct Example {}

#[gdnative::methods]
impl Example {
    pub fn new(_owner: TRef<Node>) -> Self {
        Example {}
    }
}
```

Gives the following error
```
error[E0308]: mismatched types
   --> src\handle\game_handle.rs:105:10
    |
105 | #[derive(gdnative::NativeClass)]
    |          ^^^^^^^^^^^^^^^^^^^^^ expected struct `gdnative::TRef`, found `&gdnative::prelude::Node`
    |
    = note: expected struct `gdnative::TRef<'_, gdnative::prelude::Node>`
            found reference `&gdnative::prelude::Node`
```

TRef is accepted as type for owner in exported methods other then new.
This compiles:
```rust
#[derive(gdnative::NativeClass)]
#[inherit(gdnative::prelude::Node)]
pub struct Example {}

#[gdnative::methods]
impl Example {
    pub fn new(_owner: &Node) -> Self {
        Example {}
    }

    #[export]
    pub fn test(&self, _owner: TRef<Node>) {
        //..
    }
}
```

--- a/gdnative-core/src/nativescript/class.rs
+++ b/gdnative-core/src/nativescript/class.rs
@@ -63,7 +63,7 @@ pub trait NativeClass: Sized + 'static {
     ///
     /// This function has a reference to the owner object as a parameter, which can be used to
     /// set state on the owner upon creation or to query values
-    fn init(owner: &Self::Base) -> Self;
+    fn init(owner: TRef<'_, Self::Base, Shared>) -> Self;
 
     /// Register any exported properties to Godot.
     #[inline]

--- a/gdnative-core/src/nativescript/init.rs
+++ b/gdnative-core/src/nativescript/init.rs
@@ -123,7 +123,7 @@ impl InitHandle {
                     };
 
                     let val = match panic::catch_unwind(AssertUnwindSafe(|| {
-                        C::init(C::Base::cast_ref(owner))
+                        C::init(TRef::new(C::Base::cast_ref(owner)))
                     })) {
                         Ok(val) => val,
                         Err(_) => {

--- a/gdnative-core/src/nativescript/init/property/accessor.rs
+++ b/gdnative-core/src/nativescript/init/property/accessor.rs
@@ -71,7 +71,7 @@ pub struct Mut;
 /// Helper trait for setters, generic over `self` argument mutability.
 pub trait MapSet<C: NativeClass, F, T> {
     type Err: Debug;
-    fn map_set(user_data: &C::UserData, op: &F, owner: &C::Base, value: T)
+    fn map_set(user_data: &C::UserData, op: &F, owner: TRef<C::Base>, value: T)
         -> Result<(), Self::Err>;
 }
 
@@ -79,14 +79,14 @@ impl<C, F, T> MapSet<C, F, T> for Shr
 where
     C: NativeClass,
     C::UserData: Map,
-    F: 'static + Fn(&C, &C::Base, T),
+    F: 'static + Fn(&C, TRef<C::Base>, T),
 {
     type Err = <C::UserData as Map>::Err;
     #[inline]
     fn map_set(
         user_data: &C::UserData,
         op: &F,
-        owner: &C::Base,
+        owner: TRef<C::Base>,
         value: T,
     ) -> Result<(), Self::Err> {
         user_data.map(|rust_ty| op(rust_ty, owner, value))
@@ -97,14 +97,14 @@ impl<C, F, T> MapSet<C, F, T> for Mut
 where
     C: NativeClass,
     C::UserData: MapMut,
-    F: 'static + Fn(&mut C, &C::Base, T),
+    F: 'static + Fn(&mut C, TRef<C::Base>, T),
 {
     type Err = <C::UserData as MapMut>::Err;
     #[inline]
     fn map_set(
         user_data: &C::UserData,
         op: &F,
-        owner: &C::Base,
+        owner: TRef<C::Base>,
         value: T,
     ) -> Result<(), Self::Err> {
         user_data.map_mut(|rust_ty| op(rust_ty, owner, value))
@@ -119,7 +119,7 @@ pub struct Ref;
 /// Helper trait for setters, generic over `self` argument mutability and return kind.
 pub trait MapGet<C: NativeClass, F, T> {
     type Err: Debug;
-    fn map_get(user_data: &C::UserData, op: &F, owner: &C::Base) -> Result<Variant, Self::Err>;
+    fn map_get(user_data: &C::UserData, op: &F, owner: TRef<C::Base>) -> Result<Variant, Self::Err>;
 }
 
 impl<C, F, T> MapGet<C, F, T> for (Shr, Owned)
@@ -127,11 +127,11 @@ where
     C: NativeClass,
     C::UserData: Map,
     T: ToVariant,
-    F: 'static + Fn(&C, &C::Base) -> T,
+    F: 'static + Fn(&C, TRef<C::Base>) -> T,
 {
     type Err = <C::UserData as Map>::Err;
     #[inline]
-    fn map_get(user_data: &C::UserData, op: &F, owner: &C::Base) -> Result<Variant, Self::Err> {
+    fn map_get(user_data: &C::UserData, op: &F, owner: TRef<C::Base>) -> Result<Variant, Self::Err> {
         user_data.map(|rust_ty| op(rust_ty, owner).to_variant())
     }
 }
@@ -141,11 +141,11 @@ where
     C: NativeClass,
     C::UserData: Map,
     T: ToVariant,
-    F: 'static + for<'r> Fn(&'r C, &C::Base) -> &'r T,
+    F: 'static + for<'r> Fn(&'r C, TRef<C::Base>) -> &'r T,
 {
     type Err = <C::UserData as Map>::Err;
     #[inline]
-    fn map_get(user_data: &C::UserData, op: &F, owner: &C::Base) -> Result<Variant, Self::Err> {
+    fn map_get(user_data: &C::UserData, op: &F, owner: TRef<C::Base>) -> Result<Variant, Self::Err> {
         user_data.map(|rust_ty| op(rust_ty, owner).to_variant())
     }
 }
@@ -155,11 +155,11 @@ where
     C: NativeClass,
     C::UserData: MapMut,
     T: ToVariant,
-    F: 'static + Fn(&mut C, &C::Base) -> T,
+    F: 'static + Fn(&mut C, TRef<C::Base>) -> T,
 {
     type Err = <C::UserData as MapMut>::Err;
     #[inline]
-    fn map_get(user_data: &C::UserData, op: &F, owner: &C::Base) -> Result<Variant, Self::Err> {
+    fn map_get(user_data: &C::UserData, op: &F, owner: TRef<C::Base>) -> Result<Variant, Self::Err> {
         user_data.map_mut(|rust_ty| op(rust_ty, owner).to_variant())
     }
 }
@@ -169,11 +169,11 @@ where
     C: NativeClass,
     C::UserData: MapMut,
     T: ToVariant,
-    F: 'static + for<'r> Fn(&'r mut C, &C::Base) -> &'r T,
+    F: 'static + for<'r> Fn(&'r mut C, TRef<C::Base>) -> &'r T,
 {
     type Err = <C::UserData as MapMut>::Err;
     #[inline]
-    fn map_get(user_data: &C::UserData, op: &F, owner: &C::Base) -> Result<Variant, Self::Err> {
+    fn map_get(user_data: &C::UserData, op: &F, owner: TRef<C::Base>) -> Result<Variant, Self::Err> {
         user_data.map_mut(|rust_ty| op(rust_ty, owner).to_variant())
     }
 }
@@ -221,7 +221,7 @@ where
 
             let result = std::panic::catch_unwind(|| unsafe {
                 let user_data = C::UserData::clone_from_user_data_unchecked(class as *const _);
-                let owner = C::Base::cast_ref(RawObject::from_sys_ref_unchecked(this));
+                let owner = TRef::new(C::Base::cast_ref(RawObject::from_sys_ref_unchecked(this)));
                 let func = &*(method as *const F);
 
                 match T::from_variant(Variant::cast_ref(val)) {
@@ -296,7 +296,7 @@ where
 
             let result = std::panic::catch_unwind(|| unsafe {
                 let user_data = C::UserData::clone_from_user_data_unchecked(class as *const _);
-                let owner = C::Base::cast_ref(RawObject::from_sys_ref_unchecked(this));
+                let owner = TRef::new(C::Base::cast_ref(RawObject::from_sys_ref_unchecked(this)));
                 let func = &*(method as *const F);
 
                 match <(SelfArg, RetKind)>::map_get(&user_data, func, owner) {

--- a/gdnative-core/src/nativescript/init/property/accessor.rs
+++ b/gdnative-core/src/nativescript/init/property/accessor.rs
@@ -71,8 +71,12 @@ pub struct Mut;
 /// Helper trait for setters, generic over `self` argument mutability.
 pub trait MapSet<C: NativeClass, F, T> {
     type Err: Debug;
-    fn map_set(user_data: &C::UserData, op: &F, owner: TRef<C::Base>, value: T)
-        -> Result<(), Self::Err>;
+    fn map_set(
+        user_data: &C::UserData,
+        op: &F,
+        owner: TRef<C::Base>,
+        value: T,
+    ) -> Result<(), Self::Err>;
 }
 
 impl<C, F, T> MapSet<C, F, T> for Shr
@@ -119,7 +123,8 @@ pub struct Ref;
 /// Helper trait for setters, generic over `self` argument mutability and return kind.
 pub trait MapGet<C: NativeClass, F, T> {
     type Err: Debug;
-    fn map_get(user_data: &C::UserData, op: &F, owner: TRef<C::Base>) -> Result<Variant, Self::Err>;
+    fn map_get(user_data: &C::UserData, op: &F, owner: TRef<C::Base>)
+        -> Result<Variant, Self::Err>;
 }
 
 impl<C, F, T> MapGet<C, F, T> for (Shr, Owned)
@@ -131,7 +136,11 @@ where
 {
     type Err = <C::UserData as Map>::Err;
     #[inline]
-    fn map_get(user_data: &C::UserData, op: &F, owner: TRef<C::Base>) -> Result<Variant, Self::Err> {
+    fn map_get(
+        user_data: &C::UserData,
+        op: &F,
+        owner: TRef<C::Base>,
+    ) -> Result<Variant, Self::Err> {
         user_data.map(|rust_ty| op(rust_ty, owner).to_variant())
     }
 }
@@ -145,7 +154,11 @@ where
 {
     type Err = <C::UserData as Map>::Err;
     #[inline]
-    fn map_get(user_data: &C::UserData, op: &F, owner: TRef<C::Base>) -> Result<Variant, Self::Err> {
+    fn map_get(
+        user_data: &C::UserData,
+        op: &F,
+        owner: TRef<C::Base>,
+    ) -> Result<Variant, Self::Err> {
         user_data.map(|rust_ty| op(rust_ty, owner).to_variant())
     }
 }
@@ -159,7 +172,11 @@ where
 {
     type Err = <C::UserData as MapMut>::Err;
     #[inline]
-    fn map_get(user_data: &C::UserData, op: &F, owner: TRef<C::Base>) -> Result<Variant, Self::Err> {
+    fn map_get(
+        user_data: &C::UserData,
+        op: &F,
+        owner: TRef<C::Base>,
+    ) -> Result<Variant, Self::Err> {
         user_data.map_mut(|rust_ty| op(rust_ty, owner).to_variant())
     }
 }
@@ -173,7 +190,11 @@ where
 {
     type Err = <C::UserData as MapMut>::Err;
     #[inline]
-    fn map_get(user_data: &C::UserData, op: &F, owner: TRef<C::Base>) -> Result<Variant, Self::Err> {
+    fn map_get(
+        user_data: &C::UserData,
+        op: &F,
+        owner: TRef<C::Base>,
+    ) -> Result<Variant, Self::Err> {
         user_data.map_mut(|rust_ty| op(rust_ty, owner).to_variant())
     }
 }

--- a/gdnative-core/src/object.rs
+++ b/gdnative-core/src/object.rs
@@ -830,7 +830,7 @@ impl<'a, T: GodotObject, Access: ThreadAccess> Borrow<T> for TRef<'a, T, Access>
 }
 
 impl<'a, T: GodotObject, Access: ThreadAccess> TRef<'a, T, Access> {
-    fn new(obj: &'a T) -> Self {
+    pub(crate) fn new(obj: &'a T) -> Self {
         TRef {
             obj,
             _marker: PhantomData,

--- a/gdnative-derive/src/native_script.rs
+++ b/gdnative-derive/src/native_script.rs
@@ -58,8 +58,8 @@ pub(crate) fn derive_native_class(input: TokenStream) -> TokenStream {
                     #name_str
                 }
 
-                fn init(owner: &Self::Base) -> Self {
-                    Self::new(owner)
+                fn init(owner: ::gdnative::TRef<Self::Base>) -> Self {
+                    Self::new(::gdnative::nativescript::OwnerArg::from_safe_ref(owner))
                 }
 
                 fn register_properties(builder: &::gdnative::nativescript::init::ClassBuilder<Self>) {

--- a/test/src/lib.rs
+++ b/test/src/lib.rs
@@ -101,32 +101,24 @@ fn test_underscore_method_binding() -> bool {
     ok
 }
 
+#[derive(NativeClass)]
+#[inherit(Reference)]
 struct Foo(i64);
 
-impl NativeClass for Foo {
-    type Base = Reference;
-    type UserData = user_data::ArcData<Foo>;
-    fn class_name() -> &'static str {
-        "Foo"
-    }
-    fn init(_owner: &Reference) -> Foo {
+impl Foo {
+    fn new(_owner: TRef<Reference>) -> Foo {
         Foo(42)
     }
-    fn register_properties(_builder: &ClassBuilder<Self>) {}
 }
 
+#[derive(NativeClass)]
+#[inherit(Reference)]
 struct NotFoo;
 
-impl NativeClass for NotFoo {
-    type Base = Reference;
-    type UserData = user_data::ArcData<NotFoo>;
-    fn class_name() -> &'static str {
-        "NotFoo"
-    }
-    fn init(_owner: &Reference) -> NotFoo {
+impl NotFoo {
+    fn new(_owner: &Reference) -> NotFoo {
         NotFoo
     }
-    fn register_properties(_builder: &ClassBuilder<Self>) {}
 }
 
 #[methods]

--- a/test/src/test_free_ub.rs
+++ b/test/src/test_free_ub.rs
@@ -14,21 +14,15 @@ pub(crate) fn register(handle: InitHandle) {
     handle.add_class::<Bar>();
 }
 
+#[derive(NativeClass)]
+#[inherit(Node)]
 struct Bar(i64, Option<Arc<AtomicUsize>>);
 
-impl NativeClass for Bar {
-    type Base = Node;
-    type UserData = user_data::RwLockData<Bar>;
-    fn class_name() -> &'static str {
-        "Bar"
-    }
-    fn init(_owner: &Node) -> Bar {
+impl Bar {
+    fn new(_owner: TRef<Node>) -> Self {
         Bar(42, None)
     }
-    fn register_properties(_builder: &ClassBuilder<Self>) {}
-}
 
-impl Bar {
     fn set_drop_counter(&mut self, counter: Arc<AtomicUsize>) {
         self.1 = Some(counter);
     }

--- a/test/src/test_register.rs
+++ b/test/src/test_register.rs
@@ -67,12 +67,12 @@ impl NativeClass for RegisterProperty {
 #[methods]
 impl RegisterProperty {
     #[export]
-    fn set_value(&mut self, _owner: &Reference, value: i64) {
+    fn set_value(&mut self, _owner: TRef<Reference>, value: i64) {
         self.value = value;
     }
 
     #[export]
-    fn get_value(&self, _owner: &Reference) -> i64 {
+    fn get_value(&self, _owner: TRef<Reference>) -> i64 {
         self.value
     }
 }

--- a/test/src/test_register.rs
+++ b/test/src/test_register.rs
@@ -22,7 +22,7 @@ impl NativeClass for RegisterSignal {
     fn class_name() -> &'static str {
         "RegisterSignal"
     }
-    fn init(_owner: &Reference) -> RegisterSignal {
+    fn init(_owner: TRef<Reference>) -> RegisterSignal {
         RegisterSignal
     }
     fn register_properties(builder: &ClassBuilder<Self>) {
@@ -51,7 +51,7 @@ impl NativeClass for RegisterProperty {
     fn class_name() -> &'static str {
         "RegisterProperty"
     }
-    fn init(_owner: &Reference) -> RegisterProperty {
+    fn init(_owner: TRef<Reference>) -> RegisterProperty {
         RegisterProperty { value: 42 }
     }
     fn register_properties(builder: &ClassBuilder<Self>) {

--- a/test/src/test_return_leak.rs
+++ b/test/src/test_return_leak.rs
@@ -31,7 +31,7 @@ impl NativeClass for Probe {
         "ReturnLeakProbe"
     }
 
-    fn init(_owner: &api::AnimationNodeAdd2) -> Probe {
+    fn init(_owner: TRef<api::AnimationNodeAdd2>) -> Probe {
         Probe { drop_count: None }
     }
 

--- a/test/src/test_variant_call_args.rs
+++ b/test/src/test_variant_call_args.rs
@@ -20,7 +20,7 @@ impl NativeClass for VariantCallArgs {
     fn class_name() -> &'static str {
         "VariantCallArgs"
     }
-    fn init(_owner: &Reference) -> VariantCallArgs {
+    fn init(_owner: TRef<Reference>) -> VariantCallArgs {
         VariantCallArgs
     }
     fn register_properties(_builder: &ClassBuilder<Self>) {}

